wavefront/resource_ingestion_policy.go

```diff
@@ -35,36 +51,126 @@ func resourceIngestionPolicy() *schema.Resource {
 				Type:     schema.TypeString,
 				Required: true,
 			},
+			ipAccountsKey: {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			ipGroupsKey: {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			ipSourcesKey: {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			ipNamespacesKey: {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			ipTagsKey: {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: ingestionPolicyTagSchema,
+				},
+			},
 		},
 	}
 }
 
-func resourceIngestionPolicyCreate(
-	d *schema.ResourceData, meta interface{}) error {
-	client := meta.(*wavefrontClient).client.IngestionPolicies()
+func ingestionPolicyTagSchema() map[string]*schema.Schema {
+	return map[string]*schema.Schema{
+		ipKeyKey: {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+		ipValueKey: {
+			Type:     schema.TypeString,
+			Required: true,
+		},
+	}
+}
+
+func flattenIngestionPolicyAccountIDs(accounts []wavefront.IngestionPolicyAccount) []string {
+	var tfMaps []string
+	if accounts != nil && len(accounts) > 0 {
```

Comment:
I think I got that pattern from here: https://github.com/vmware/terraform-provider-wavefront/blob/master/wavefront/resource_metrics_policy.go#L274 (Is it also unnecessary here?)

Reply:
In the link you provided, the scenario is a little different. raw is an interface{} type, so raw can either be nil (have nothing) or it can contain a value of any type.  The raw.([]interface{}) expression is called a type assertion. It gets the underlying []interface{} concrete value out of raw. If raw happens to be nil (contains nothing) the type assertion will panic, and the program will crash.  So checking that raw != nil makes sense, but checking that len(raw.([]interface{})) is bigger than zero is not necessary.  In your code, accounts is not an interface but rather a slice.  For all practical purposes a nil slice has length 0, and you can append to a nil slice just like you can append to a zero length slice. All this may seem confusing because the meaning of nil in go changes depending on the context. For interface nil means contains no value. For slices and maps nil means empty. But for maps, you can't insert new key value pairs into a nil map.

Reply:
> ...the type assertion will panic...

I see, interesting.

I know what an interface means in a computer science sense but, I get the impression it has additional meaning in Go.

> ...maps nil means empty. But for maps, you can't insert new key value pairs into a nil map.

Also interesting 👍